File: server/lib/models/agent_run.dart
class AgentRun {
  AgentRun({
    required this.id,
    required this.workItemId,
    required this.model,
    required this.status,
    required this.summary,
    required this.startedAt,
    required this.finishedAt,
    required this.maxSteps,
    required this.error,
  });

  final String id;
  final String workItemId;
  final String model;
  final String status;
  final String? summary;
  final DateTime startedAt;
  final DateTime? finishedAt;
  final int maxSteps;
  final String? error;

  factory AgentRun.fromRow(Map<String, Object?> r) => AgentRun(
        id: r['id']!.toString(),
        workItemId: r['work_item_id']!.toString(),
        model: r['model'] as String,
        status: r['status'] as String,
        summary: r['summary'] as String?,
        startedAt: DateTime.parse(r['started_at']!.toString()).toUtc(),
        finishedAt: r['finished_at'] == null
            ? null
            : DateTime.parse(r['finished_at']!.toString()).toUtc(),
        maxSteps: (r['max_steps'] as num).toInt(),
        error: r['error'] as String?,
      );

  Map<String, Object?> toJson() => {
        'id': id,
        'workItemId': workItemId,
        'model': model,
        'status': status,
        'summary': summary,
        'startedAt': startedAt.toIso8601String(),
        'finishedAt': finishedAt?.toIso8601String(),
        'maxSteps': maxSteps,
        'error': error,
      };
}

class PlanStep {
  PlanStep({
    required this.id,
    required this.runId,
    required this.ordinal,
    required this.title,
    required this.rationale,
    required this.status,
    required this.result,
    required this.startedAt,
    required this.finishedAt,
  });

  final String id;
  final String runId;
  final int ordinal;
  final String title;
  final String? rationale;
  final String status;
  final String? result;
  final DateTime? startedAt;
  final DateTime? finishedAt;

  factory PlanStep.fromRow(Map<String, Object?> r) => PlanStep(
        id: r['id']!.toString(),
        runId: r['run_id']!.toString(),
        ordinal: (r['ordinal'] as num).toInt(),
        title: r['title'] as String,
        rationale: r['rationale'] as String?,
        status: r['status'] as String,
        result: r['result'] as String?,
        startedAt: r['started_at'] == null
            ? null
            : DateTime.parse(r['started_at']!.toString()).toUtc(),
        finishedAt: r['finished_at'] == null
            ? null
            : DateTime.parse(r['finished_at']!.toString()).toUtc(),
      );

  Map<String, Object?> toJson() => {
        'id': id,
        'runId': runId,
        'ordinal': ordinal,
        'title': title,
        'rationale': rationale,
        'status': status,
        'result': result,
        'startedAt': startedAt?.toIso8601String(),
        'finishedAt': finishedAt?.toIso8601String(),
      };
}

class AgentMessage {
  AgentMessage({
    required this.id,
    required this.runId,
    required this.stepId,
    required this.role,
    required this.content,
    required this.toolName,
    required this.createdAt,
  });

  final String id;
  final String runId;
  final String? stepId;
  final String role;
  final String content;
  final String? toolName;
  final DateTime createdAt;

  factory AgentMessage.fromRow(Map<String, Object?> r) => AgentMessage(
        id: r['id']!.toString(),
        runId: r['run_id']!.toString(),
        stepId: r['step_id']?.toString(),
        role: r['role'] as String,
        content: r['content'] as String,
        toolName: r['tool_name'] as String?,
        createdAt: DateTime.parse(r['created_at']!.toString()).toUtc(),
      );

  Map<String, Object?> toJson() => {
        'id': id,
        'runId': runId,
        'stepId': stepId,
        'role': role,
        'content': content,
        'toolName': toolName,
        'createdAt': createdAt.toIso8601String(),
      };
}

File: server/app/models/agent_run.py
from __future__ import annotations

from datetime import datetime
from typing import Any

from pydantic import BaseModel, ConfigDict, Field

from .work_item import _parse_dt


class AgentRun(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    id: str
    work_item_id: str = Field(alias="workItemId")
    model: str
    status: str
    summary: str | None = None
    started_at: datetime = Field(alias="startedAt")
    finished_at: datetime | None = Field(default=None, alias="finishedAt")
    max_steps: int = Field(alias="maxSteps")
    error: str | None = None

    @classmethod
    def from_row(cls, row: dict[str, Any]) -> AgentRun:
        return cls(
            id=str(row["id"]),
            workItemId=str(row["work_item_id"]),
            model=row["model"],
            status=row["status"],
            summary=row.get("summary"),
            startedAt=_parse_dt(row["started_at"]),
            finishedAt=_parse_dt(row["finished_at"]) if row.get("finished_at") else None,
            maxSteps=int(row["max_steps"]),
            error=row.get("error"),
        )


class PlanStep(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    id: str
    run_id: str = Field(alias="runId")
    ordinal: int
    title: str
    rationale: str | None = None
    status: str
    result: str | None = None
    started_at: datetime | None = Field(default=None, alias="startedAt")
    finished_at: datetime | None = Field(default=None, alias="finishedAt")

    @classmethod
    def from_row(cls, row: dict[str, Any]) -> PlanStep:
        return cls(
            id=str(row["id"]),
            runId=str(row["run_id"]),
            ordinal=int(row["ordinal"]),
            title=row["title"],
            rationale=row.get("rationale"),
            status=row["status"],
            result=row.get("result"),
            startedAt=_parse_dt(row["started_at"]) if row.get("started_at") else None,
            finishedAt=_parse_dt(row["finished_at"]) if row.get("finished_at") else None,
        )


class AgentMessage(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    id: str
    run_id: str = Field(alias="runId")
    step_id: str | None = Field(default=None, alias="stepId")
    role: str
    content: str
    tool_name: str | None = Field(default=None, alias="toolName")
    created_at: datetime = Field(alias="createdAt")

    @classmethod
    def from_row(cls, row: dict[str, Any]) -> AgentMessage:
        return cls(
            id=str(row["id"]),
            runId=str(row["run_id"]),
            stepId=str(row["step_id"]) if row.get("step_id") else None,
            role=row["role"],
            content=row["content"],
            toolName=row.get("tool_name"),
            createdAt=_parse_dt(row["created_at"]),
        )
